File: ios/wby/wby/Models/WeatherSymbols.swift
import CoreLocation
import Foundation

enum WeatherSymbols {
    static func primarySymbol(from weather: WeatherResponse?) -> String? {
        weather?.hourlyForecast.first?.symbol ?? weather?.dailyForecast.first?.symbol
    }

    static func scene(for weather: WeatherResponse?) -> WeatherScene {
        WeatherScene.from(symbolCode: primarySymbol(from: weather))
    }

    static func scene(
        for weather: WeatherResponse?,
        coordinate: CLLocationCoordinate2D,
        date: Date,
        timeZone: TimeZone,
        elevationMeters: Double
    ) -> WeatherScene {
        WeatherScene.from(symbolCode:
            nightAdjusted(
                primarySymbol(from: weather),
                coordinate: coordinate,
                at: date,
                timeZone: timeZone,
                elevationMeters: elevationMeters
            )
        )
    }

    static func nightAdjusted(
        _ symbolCode: String?,
        coordinate: CLLocationCoordinate2D,
        at date: Date,
        timeZone: TimeZone,
        elevationMeters: Double
    ) -> String? {
        guard let code = symbolCode.flatMap(Int.init), code < 100 else { return symbolCode }
        let isNight = SunriseCard.isNight(
            coordinate: coordinate,
            date: date,
            timeZone: timeZone,
            elevationMeters: elevationMeters
        )
        return isNight ? String(code + 100) : symbolCode
    }

    static func conditionDescription(from weather: WeatherResponse?) -> String? {
        conditionDescription(from: primarySymbol(from: weather))
    }

    static func conditionDescription(from symbolCode: String?) -> String? {
        guard let code = symbolCode.flatMap(Int.init) else { return nil }
        let n = code >= 100 ? code - 100 : code
        switch n {
        case 1: return "Clear"
        case 2: return "Mostly Clear"
        case 4: return "Partly Cloudy"
        case 6: return "Mostly Cloudy"
        case 7: return "Cloudy"
        case 9: return "Fog"
        case 11: return "Drizzle"
        case 14: return "Freezing Drizzle"
        case 17: return "Freezing Rain"
        case 21: return "Isolated Showers"
        case 24: return "Scattered Showers"
        case 27: return "Showers"
        case 31: return "Isolated Light Showers"
        case 32: return "Isolated Showers"
        case 33: return "Isolated Heavy Showers"
        case 34: return "Scattered Light Showers"
        case 35: return "Scattered Showers"
        case 36: return "Scattered Heavy Showers"
        case 37: return "Light Rain"
        case 38: return "Rain"
        case 39: return "Heavy Rain"
        case 41: return "Isolated Light Sleet Showers"
        case 42: return "Isolated Sleet Showers"
        case 43: return "Isolated Heavy Sleet Showers"
        case 44: return "Scattered Light Sleet Showers"
        case 45: return "Scattered Sleet Showers"
        case 46: return "Scattered Heavy Sleet Showers"
        case 47: return "Light Sleet"
        case 48: return "Sleet"
        case 49: return "Heavy Sleet"
        case 51: return "Isolated Light Snow Showers"
        case 52: return "Isolated Snow Showers"
        case 53: return "Isolated Heavy Snow Showers"
        case 54: return "Scattered Light Snow Showers"
        case 55: return "Scattered Snow Showers"
        case 56: return "Scattered Heavy Snow Showers"
        case 57: return "Light Snow"
        case 58: return "Snow"
        case 59: return "Heavy Snow"
        case 61: return "Isolated Hail Showers"
        case 64: return "Scattered Hail Showers"
        case 67: return "Hail Showers"
        case 71: return "Isolated Thundershowers"
        case 74: return "Scattered Thundershowers"
        case 77: return "Thundershowers"
        default: return nil
        }
    }
}

File: web/src/lib/weatherSymbols.ts
/**
 * FMI SmartSymbol weather codes, as returned in the `symbol` field of hourly
 * and daily forecasts. Codes >= 100 are night variants of `code - 100`.
 *
 * Labels and the night-asset fallback mirror the official FMI weather app.
 */

export type SymbolTheme = "light" | "dark";

export const SYMBOL_DESCRIPTIONS: Readonly<Record<number, string>> = {
  1: "Clear",
  2: "Mostly clear",
  4: "Partly cloudy",
  6: "Mostly cloudy",
  7: "Cloudy",
  9: "Fog",
  11: "Drizzle",
  14: "Freezing drizzle",
  17: "Freezing rain",
  21: "Isolated showers",
  24: "Scattered showers",
  27: "Showers",
  31: "Isolated light showers",
  32: "Isolated moderate showers",
  33: "Isolated heavy showers",
  34: "Scattered light showers",
  35: "Scattered moderate showers",
  36: "Scattered heavy showers",
  37: "Light rain",
  38: "Moderate rain",
  39: "Heavy rain",
  41: "Isolated light sleet showers",
  42: "Isolated moderate sleet showers",
  43: "Isolated heavy sleet showers",
  44: "Scattered light sleet showers",
  45: "Scattered moderate sleet showers",
  46: "Scattered heavy sleet showers",
  47: "Light sleet",
  48: "Moderate sleet",
  49: "Heavy sleet",
  51: "Isolated light snow showers",
  52: "Isolated moderate snow showers",
  53: "Isolated heavy snow showers",
  54: "Scattered light snow showers",
  55: "Scattered moderate snow showers",
  56: "Scattered heavy snow showers",
  57: "Light snowfall",
  58: "Moderate snowfall",
  59: "Heavy snowfall",
  61: "Isolated hail showers",
  64: "Scattered hail showers",
  67: "Hail showers",
  71: "Isolated thundershowers",
  74: "Scattered thundershowers",
  77: "Thundershowers",
};

/** Codes that have dedicated night artwork (`src/assets/symbols/<theme>/1xx.svg`). */
const NIGHT_ASSET_CODES: ReadonlySet<number> = new Set([
  1, 2, 4, 6, 21, 24, 31, 32, 33, 34, 35, 36, 41, 42, 43, 44, 45, 46, 51, 52,
  53, 54, 55, 56, 61, 64, 71, 74,
]);

export type ParsedSymbol = {
  /** Day code (night offset removed). */
  code: number;
  night: boolean;
};

export function parseSymbolCode(raw: string | null | undefined): ParsedSymbol | null {
  if (raw == null) return null;
  const value = Number.parseInt(raw, 10);
  if (!Number.isFinite(value) || value < 0) return null;
  const night = value >= 100;
  const code = night ? value - 100 : value;
  if (!(code in SYMBOL_DESCRIPTIONS)) return null;
  return { code, night };
}

export function symbolDescription(raw: string | null | undefined): string | null {
  const parsed = parseSymbolCode(raw);
  return parsed ? SYMBOL_DESCRIPTIONS[parsed.code] : null;
}

/**
 * Relative file under `src/assets/symbols/` for the given code and theme,
 * falling back to the day artwork for night codes without a night icon.
 */
export function symbolAssetFile(
  raw: string | null | undefined,
  theme: SymbolTheme,
): string | null {
  const parsed = parseSymbolCode(raw);
  if (!parsed) return null;
  const file =
    parsed.night && NIGHT_ASSET_CODES.has(parsed.code) ? parsed.code + 100 : parsed.code;
  return `${theme}/${file}.svg`;
}
